src/client/datascience/kernel-launcher/kernelFinder.ts

```
@@ -0,0 +1,232 @@
+        const notebookInterpreter = isResource(interpreterUri) ? undefined : interpreterUri;
+
+        if (kernelName) {
+            let kernelSpec = await this.findCachePath(kernelName);
+
+            if (kernelSpec) {
+                return kernelSpec;
+            }
+
+            if (!notebookInterpreter) {
+                kernelSpec = await this.getKernelSpecFromActiveInterpreter(resource, kernelName);
+            }
+
+            if (kernelSpec) {
+                return kernelSpec;
+            }
+
+            const kernelSearches = [
+                this.interpreterService.getInterpreters(resource).then((interpreters) => {
+                    const interpreterPaths = interpreters.map((interp) => interp.path);
+                    return this.findInterpreterPath(interpreterPaths, kernelName);
+                }),
+                this.findDiskPath(kernelName),
+                createPromiseFromCancellation({
+                    defaultValue: undefined,
+                    cancelAction: 'resolve',
+                    token: token
+                })
+            ];
+
+            const result = await Promise.race(kernelSearches);
+            return result ? result : this.getDefaultKernelSpec(resource);
+        }
+
+        return this.getDefaultKernelSpec(resource);
+    }
+
+    private async getKernelSpecFromActiveInterpreter(
+        resource: Resource,
+        kernelName?: string
+    ): Promise<IJupyterKernelSpec | undefined> {
+        this.activeInterpreter = await this.interpreterService.getActiveInterpreter(resource);
+
+        if (this.activeInterpreter) {
+            return this.getKernelSpec(path.join(this.activeInterpreter.path, 'share/jupyter/kernels/'), kernelName);
+        }
+    }
+
+    private async getKernelSpec(kernelPath: string, kernelName?: string): Promise<IJupyterKernelSpec | undefined> {
+        const pathExists = await this.file.directoryExists(kernelPath);
```

We can avoid this.
Else its another IO that slows down searching.
I.e. if `this.file.readFile` fails, then we can ignore it and exit (it would fail if file is not found, which automatically means directory doesn't exist, hence one less disc io).